ENH - Update links styling

trallard opened the pull request:
This PR introduces changes to the theme's link styling to meet [WCAG SC 1.4.1](https://www.w3.org/WAI/WCAG21/Understanding/use-of-color.html)

Particularly by implementing the [technique G182](https://www.w3.org/WAI/WCAG21/Techniques/general/G182) - which introduces a redundant visual cue for folks (i.e., underline, sidebar vertical borders)

Note this PR does not fix focus visible affordances, as this will be done in a subsequent PR (this includes all focus states: links, dropdowns, buttons, etc., as I would like @smeragoel to have a look at the proposed implementation we have in the Figma file).

🤔 Question: does anyone know where I have to change the behavior of the `navbar` links? I have tried a couple of places/approaches and cannot get the `hover` and `active` underline visible for my life.

closes #835

drammock commented:
> does anyone know where I have to change the behavior of the `navbar` links?

weird, I see potentially 2 places, maybe that's the problem?

https://github.com/pydata/pydata-sphinx-theme/blob/cb23a7a7f5414d3cb3c7d9fe7c422757e4e21216/src/pydata_sphinx_theme/assets/styles/sections/_header.scss#L74-L97

https://github.com/pydata/pydata-sphinx-theme/blob/cb23a7a7f5414d3cb3c7d9fe7c422757e4e21216/src/pydata_sphinx_theme/assets/styles/components/_navbar-links.scss#L15-L43

or do you mean the icon links? Those are in this file:

https://github.com/pydata/pydata-sphinx-theme/blob/cb23a7a7f5414d3cb3c7d9fe7c422757e4e21216/src/pydata_sphinx_theme/assets/styles/components/_icon-links.scss#L5

trallard commented:
The two first ones are the ones I am after. I might have only changed in one place and did not notice the second file. Thanks.


I was planning to leave the icons until a later iteration but it might be best to have at least the default and hover states updated while I re-evaluate focus states.

trallard commented:
Thank you @drammock you pointed me in the right direction. 

This should be ready for review.

drammock commented on src/pydata_sphinx_theme/assets/styles/components/_navbar-links.scss:
```diff
@@ -21,24 +21,22 @@
     }
 
-    // Active page is always highlighted
+    // Active page is always underlined - using the same underline style as hover
+    // with the primary colour
```
I'm not seeing this behavior on the PR build ("user guide" has no underline):

![Screenshot 2023-06-26 at 12-00-34 Accessibility — PyData Theme 0 13 4dev0 documentation](https://github.com/pydata/pydata-sphinx-theme/assets/1810515/c42b2983-750d-4c6c-b3be-10265ef5b318)

drammock replied:
I think we're talking about different things here. My understanding of the code comment was that the active section name would be teal with an underline *even when not hovering on it* --- my screenshot shows that it is teal but lacks an underline.  

Your screenshot seems to show what happens *when hovering* on the active section name --- secondary color with thick underline (which is expected and correct, but is not the problem I was trying to point out).

trallard replied:
aaaaah yes got confused - I noticed the underline has gone missing, and it is in my mind to fix it.
This is another example of having to fish changes in two places.

drammock commented on src/pydata_sphinx_theme/assets/styles/components/_search.scss:
```diff
@@ -65,7 +65,7 @@
 
   &:hover {
-    color: var(--pst-color-primary);
+    color: var(--pst-color-secondary);
```
equivalent result, but easier-to-understand source?
```suggestion
    color: var(--pst-color-link-hover);
```

drammock commented on src/pydata_sphinx_theme/assets/styles/sections/_sidebar-primary.scss:
```diff
@@ -207,11 +201,12 @@ nav.bd-links {
     }
   }
 
+  // Hover should still work on top of the active link
```
this works on the text, but not on the div-edge color:

![Screenshot_2023-06-26_12-06-13](https://github.com/pydata/pydata-sphinx-theme/assets/1810515/7fb23838-29c0-4e2a-9b98-5ba269f3b8d9)

you can't see my cursor in the screenshot, but (perhaps obviously) it's on the "theme variables and CSS" TOC item

trallard replied:
That is intentional - in menus, the hover, active and focus states should not interfere with the current state (i.e. a menu item could be focused with a keyboard and retain its location/current indicator simultaneously). So in the sidebar:

* The teal bar [indicates the current location within the sidebar](https://www.w3.org/TR/WCAG20-TECHS/G128)
* The hover should be displayed alongside the current state (n[ot override it](https://www.w3.org/WAI/tutorials/menus/styling/#hover-and-focus-states) 

But as I write this, I realise that this is directly violating the `not rely on colour` criteria - so I might need to think about how to convey this instead (in terms of the thickness of the sidebar or similar)

trallard replied:
Just pushed an update - the `hover` style (thick underline) is applied on top of the `current` one.

drammock replied:
ok I'm happy with that solution!

drammock reviewed (commented):
thanks for this @trallard, and sorry for the slow review.

In addition to the detailed comments below, I wonder if you could offer an opinion about whether we should merge (parts of?) the two source documents (`styles/components/_navbar-links` and `styles/sections/_header`) that were the cause of some frustration earlier when trying to change the navbar styles. WDYT?

trallard commented on docs/user_guide/accessibility.md:
yes I just changed the file from rst to mystMD

trallard commented on src/pydata_sphinx_theme/assets/styles/abstracts/_accessibility.scss:
Note: I split what was in `mixins` into multiple "helpers" to keep things separate rather than dumping everything that fits that function into a kitchen sink file (`mixins.scss`)

trallard commented:
@smeragoel - I had not thought about how to indicate the `hover` state on dropdowns within the navbar. Right now, as [@drammock  pointed out, there is no visible hover on the dropdown until this is expanded](https://github.com/pydata/pydata-sphinx-theme/pull/1353#discussion_r1242489833) (see video)

https://github.com/pydata/pydata-sphinx-theme/assets/23552331/488bea28-edd5-4cbe-a117-13b4bdf56a29

Do you have any recommendations on how to fix this (I had not considered this case as I left dropdown styling until later as we review the rest of the interactive elements)

trallard commented:
I can look into those icons @choldgraf - I am not 100% happy with those either as now they look odd (different from each other).

> Not sure if it's related to this change or not, but I noticed a kind-of distracting "UI hover effect" that happens on mouse click-down events. It very temporarily highlights a section as active until the mouse is un-pressed, which feels like it may not be intended and is kind-of distracting:

Definitely because of the PR - essentially, buttons and links should have the following states `default, hover, focus, active` and, in some cases `current` (like nav bars and TOCs ref: https://www.w3.org/WAI/tutorials/menus/styling/#currentnav)

An `active` state denotes the menu item that was clicked, tapped, or selected through the keyboard (a.k.a activated). Right now, that `active` state looks the same as the `hover` state. 
The state's purpose is to indicate to the user that an interactive element has been activated (and is good to identify unintended activations). Many websites and apps remove this `active` styling but then it also removes valuable visible cues for users.
I can try and make this more subtle or less distracting but I would rather not remove it for the above reasons.

drammock commented:
> An `active` state denotes the menu item that was clicked, tapped, or selected through the keyboard (a.k.a activated). Right now, that `active` state looks the same as the `hover` state.

might this be a good situation to use our "accent" color?  I know we need to do something else too in addition to color (maybe a thin all-around border?)  Just throwing out ideas, feel free to ignore if you think of something better :)

trallard commented:
Righto folks, I think I hunted and fixed most of the links - it ended up being more than I though. This PR should be okay to merge and might need some fixes later. 
There are, however, some things to note;

> The underline spans gaps when there is a fontawesome glyph within the link (e.g. on the right sidebar). Do you know of any way to prevent that from happening without de-linking the glyph?

I could not find a way to do this (I will need to open an issue and try and fix this later)

* I added a "temporary" focus indicator (many items were missing a focus indicator altogether). It is in our accent colour, but I am not sure about this, so I will wait for @smeragoel recommendations and fix this and the `active` states in another PR

* Also - visited links are now in our accent colour. I remember something about folks feeling unsure about too much pink. So shall we keep this in pink, or shall I use a different colour (perhaps a darker shade of the primary or secondary colour instead?)

We are also planning to do a keyboard navigation audit in the upcoming couple of weeks so that should be useful to identify more focus indicators missing or keyboard traps.

drammock reviewed (approved):
looks great, +1 for merge and addressing remaining bits in subsequent PRs. Two things that stand out to me that we might want to discuss more and possibly change later:

1. On e.g. [the user guide page](https://pydata-sphinx-theme--1353.org.readthedocs.build/en/1353/user_guide/index.html) the in-page TOC tree looks quite messy to me:

   ![Screenshot 2023-06-30 at 12-13-01 User Guide — PyData Theme 0 13 4dev0 documentation](https://github.com/pydata/pydata-sphinx-theme/assets/1810515/07fd6ce0-6833-47ee-8f08-9c249a5c6600)

   I've never much liked having the TOCtree as part of the main content, so one solution is to just remove it and let people navigate with the prev/next buttons or the left sidebar links. Another solution is some kind of formatting change: either special-casing these links to have no underline (or a more subtle one?) or possibly adding some visual structure (like bullets or tree-view lines?) to make it seem less chaotic.  Or some other approach that I haven't thought of?

2. I'm not a huge fan of distinguishing "visited" links with a different color, and (as someone with typical color vision) I find the interleaving of teal-vs-pink distracting.  For me, if we're going to have visited links in a different color, I prefer a color that is less distinct from the "normal" link state (teal).  (that said, I think using pink for the "active" state when doing keyboard navigation is great --- that's a case where it should stand out prominently)

trallard commented:
Thanks @drammock 

1. Agree there is something that looks a bit odd from the TOCtree. It might be worth opening a separate issue to go over this and make some improvements.
2. Agree here too, the pink colour stands out a lot and I am _almost_ sure some folks will complain about it. There is no specific requirement accessibility-wise to distinguish visited and unvisited links so I suppose we can remove it or make it more subtle. Will work on this

trallard commented:
Addressed the last few comments so if y'all are happy with this it could be merged

drammock commented:
thanks @trallard!!
